src/vision/status_quo/grace_wants_to_integrate_c_api.md

```diff
@@ -0,0 +1,176 @@
+
+```rust
+let frame_stream = stream_from_camera(camera, Duration::from_millis(5));
+
+while let Some(frame) = frame_stream.next().await {
+    // process frames
+}
+
+println!("Frame stream closed.");
+```
+
+She scours existing crates, in particular looking for one way to transform the above future into a stream that can be
+executed many times. The only available option to transform a future into a series of futures is [`stream::unfold`],
+which seems to do exactly what Grace is looking for. Grace begins by adding a small intermediate type, and then plugging
+in the remaining holes:
+
+```rust
+struct StreamState {
+    camera: Camera,
+    retry_after: Duration,
+}
+
+
+
+fn stream_from_camera(camera: Camera, retry_after: Duration) -> Unfold<Frame, ??, ??> {
+    let initial_state = StreamState { camera, retry_after };
+
+    stream::unfold(initial_state, |state| async move {
+        let frame = next_frame(&state.camera, state.retry_after).await
+        (frame, state)
+    })
+}
+```
+
+This looks like it mostly hits the mark, but Grace is left with a couple of questions for how to get the remainder of
+this building:
+
+1. What is the type that fills in the third template parameter in the return? It should be the type of the future that
+   is returned by the async closure passed into `stream::unfold`, but we don't know the type of a closure!
+2. What is the type that fills in the second template parameter of the closure in the return?
+
+Grace spends a lot of time trying to figure out how she might find those types! She asks [Barbara] what the idiomatic
+way to get around this in Rust would be. Barbara explains again how closures don't have concrete types, and that the
+only way to do this will be to use the `impl` keyword.
+
+```rust
+fn stream_from_camera(camera: Camera, retry_after: Duration) -> impl Stream<Item = Frame> {
+    // same as before
+}
+```
+
+While Grace was was on the correct path and now her team is able to write the code they want to, she realizes that
+sometimes writing the types out explicitly can be very hard. She reflects on what it would have taken to write the type
+of an equivalent function pointer in C, and slightly laments that Rust cannot express such as clearly.
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * Rust was the correct choice for the team across the board thanks to its memory safety and ownership. The
+      underlying C library was just too complex for any single programmer to be able to maintain in their head all at
+      once while also trying to accomplish other tasks.
+    * Evolving requirements meant that the team would have had to either start over in plain C, giving up a lot of the
+      safety they would gain from switching to Rust, or exploring async code in a more rigorous way.
+    * The async code is actually much simpler than writing the entire execution loop in C themselves. However, the
```

Comment:
Let me know if I'm off-base here, but it seems like the underlying moral here is that integrating Rust async code with other languages which might have different mental models can sometimes lead to unidiomatic or unsatisfying code. Is that correct?

Reply:
Sounds plausible, although I found the "sleep.await" to be .. tolerable. Maybe it'd be nice to have a future combinator or something that's like "poll every so often" just to express more clearly what's going on, though. (Maybe this exists, idk)

Reply:
I would say the assessment above is correct! Spinning on a sleep timer when using threads is actually an easy way to get race conditions in C or C++. This is why `std::condition_variable` exists, to make waiting on a certain condition across threads more clear. 

Sleeping in a future here doesn't have any of the same problems as the above, as race conditions aren't really the concern given the over-arching async model. But yes, the moral is effectively as @rylev described. Given that the wording there is much better, I think I'll go ahead and change this to say:

> The async code is actually much simpler than writing the entire execution loop in C themselves. However, the assumption that you would write the entire execution loop is baked into the underlying library which Grace's team cannot rewrite entirely from scratch. Integrating Rust async code with other languages which might have different mental models can sometimes lead to unidiomatic or unsatisfying code, even if the intent of the code in Rust is clear. 

On the topic of some kind of future combinator: this would be helpful! I don't know if that's over-abstraction though, as it seems like loop & sleep work well! One thing I did _not_ mention was that `task::sleep` is actually in `async_std::task`, not `std::task`. I think given how "basic" this is and how easily it maps back to `std::thread::sleep`, there should at least be an async sleep in `std`, so that the above example is truly independent of crate.

Reply:
See 2eedd5a1af3b923030a9e1ab15fdeb0a7451988e for the above change.